### ISSUE TITLE
Upgrade from 4.2.1409.1722 to 4.2.1507.0118 changed VerifyAll behavior

### ISSUE THREAD

**Issue by ce-smith** [moq/moq4#191]
VerifyAll() fails for setups that weren't created following the upgrade. In particular, any mock made from an interface containing a get-only property fails on not satisfying a set on that property. All failing tests **do not have setups** for the methods/properties being reported as not matched.

```
Moq.MockVerificationException: Moq.MockVerificationException: The following setups were not matched:

IDbContext m => m.Database
```

where IDbContext is defined:

```
    public interface IDbContext : IDisposable
    {
        Database Database { get; }
        // Other methods...
    }
```

Additionally, there seems to be a recursive traversal of properties that are somehow getting automatically wired up as a setup. Here, IConfigurationManager defined as:

```
    public interface IConfigurationManager
    {
        NameValueCollection AppSettings { get; }
        // Other methods...
    }
```

Any tests using a mock of type IConfigurationManager produces the following failure message:

```
Moq.MockVerificationException: Moq.MockVerificationException: The following setups were not matched:

NameValueCollection m => m.AllKeys

NameValueCollection m => m.Count

NameValueCollection m => m.Keys

NameValueCollection m => m.Count

NameValueCollection m => m.SyncRoot

NameValueCollection m => m.IsSynchronized
```

**Comment by kzu** [comment]
Could you please provide a complete and bare-bones interface definition and accompanying (single) unit test with the relevant setup which fails?

Thanks!

**Comment by cherokeejamessmith** [comment]
Hey @kzu, thanks for the reply. I went through the process of creating a stand alone solution to reproduce the problem and discovered an additional detail. The problem manifests when chaining into a subdependency on setup where that subdependency has a get-only property exposed on the interface. Here's the repo code:
```
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;

namespace MoqTest
{
	[TestClass]
	public class Tests
	{
		[TestMethod]
		public void DoSomething_NormalConditions_CallsDependency()
		{
			var mockDependency = new Mock<IDependency>();
			var service = new TestService( mockDependency.Object );
			mockDependency
				.Setup( m => m.CreateDependency().Foo() );

			service.DoSomething();

			mockDependency.VerifyAll();
		}
	}

	public class Payload
	{
	}

	public interface IDependency
	{
		ISubDependency CreateDependency();
	}

	public interface ISubDependency
	{
		void Foo();
		Payload Payload { get; }
	}

	public class TestService
	{
		private readonly IDependency _dependency;

		public TestService( IDependency dependency )
		{
			_dependency = dependency;
		}

		public void DoSomething()
		{
			_dependency.CreateDependency().Foo();
		}
	}
}
```

Here is the Test output:
```
Test method MoqTest.Tests.DoSomething_NormalConditions_CallsDependency threw exception: 

Moq.MockVerificationException: The following setups were not matched:

ISubDependency m => m.Payload


    at Moq.Mock.VerifyAll()
   at MoqTest.Tests.DoSomething_NormalConditions_CallsDependency() in TestClass.cs: line 19


```

This was a new solution in Visual Studio 2013 Update 4 targeting the .Net 4.5.2 framework. Here are the references used by the project:

```
    <Reference Include="Microsoft.VisualStudio.QualityTools.UnitTestFramework, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
    <Reference Include="Moq, Version=4.2.1507.118, Culture=neutral, PublicKeyToken=69f491c39445e920, processorArchitecture=MSIL">
      <HintPath>..\packages\Moq.4.2.1507.0118\lib\net40\Moq.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="System" />
    <Reference Include="System.Core" />
    <Reference Include="System.Xml.Linq" />
    <Reference Include="System.Data.DataSetExtensions" />
    <Reference Include="Microsoft.CSharp" />
    <Reference Include="System.Data" />
    <Reference Include="System.Xml" />
```

**Comment by ce-smith** [comment]
Any update available on this?

**Comment by kzu** [comment]
Nop.

It might be worth it to just verify what you really expect to have happened
though. The verify all behavior is much more aligned for strict mocks than
loose mocks (aka stubs).

**Comment by ce-smith** [comment]
That's disappointing since this works fine with a long range of previous versions. I'm not really interested to rewriting an entire unit test suite to adopt a new version of Moq. I hope this get addressed.

**Comment by kzu** [comment]
You can always stay with whichever version works for you.

Moq is pretty much a community driven project at this point, and it may be
that some of those contributions ended up affecting your scenario. If it's
really key for you to get the new contributions while still keeping this
behavior, it could be worth it to attempt to fix it and send a PR. We're
more that open to contributions.

Thanks

**Comment by kzu** [comment]
Marking as up-for-grabs since it doesn't seem to be a high-impact one.

**Comment by NecroFilja** [comment]
Having the same issue. About 200 tests started to failing.
It is really ugly that for 2 years there is no workaround.

And just to upgrade all packages to latest version I need to replace 200 VerifyAll on verifiable()/Verify

**Comment by stakx** [comment]
@ce-smith: I realise this was posted long ago, but I'd like to add my ¢2. Here's your unit test in slightly modified order:

> ```csharp
> var mockDependency = new Mock<IDependency>();
> mockDependency.Setup(m => m.CreateDependency().Foo());
>
> var service = new TestService(mockDependency.Object);
> service.DoSomething();
>
> mockDependency.VerifyAll();
> ```

Judging from the test, you probably assumed that the only two things that get set up are:

* `dependencyCreateDependency()`
* `createdSubDependency.Foo()`

But "recursive" setup expressions such as `m => m.CreateDependency().Foo()` actually do more, they set up all properties in the traversed objects, too. The only reason why your test used to work with version 4.2.1409.1722 is essentially because that version was affected by a regression bug: `mock.SetupAllProperties()` wouldn't actually set up all properties, but only get-set ones. Your property `ISubDependency.Payload` is read-only, so in that old version of Moq, it didn't get set up and therefore would be skipped during `VerifyAll`.

PR #137 (which was included in [Moq 4.2.1502.911](https://github.com/moq/moq4/blob/master/CHANGELOG.md#421502911-2015-02-09 "see changelog")) addressed that bug. Now, *all* properties get set up by `SetupAllProperties`, and that's why your test starts failing.

The only way how your issue could be solved would be to change "recursive" mock setups so that they only set up what's contained in the setup expression.

**Comment by stakx** [comment]
Due to the decision made in https://github.com/moq/moq4/issues/681 and the corresponding change made to `SetupAllProperties` in #684, this issue will be fixed in Moq version >4.9.0.
